### DISCUSSION TITLE
Create Titanium modules in Javascript with Hyperloop

### DISCUSSION THREAD

**Discussion by EtnaTraining** [appcelerator/titanium_mobile#12663] [Q&A]
Hello,

I have a Java library (.aar) that I need to wrap in a Titanium modules. I have already experiment with it via Hyperloop and it works fine. However I want to hide the implementation details and expose with a simpler API to be used in other Titanium projects. Moreover I would like to avoid to give the source code of the module for a possibile customer.
Could I pack the hyperloop code as a Common.js module?

Thanks

**Comment by EtnaTraining** [comment]
Today I made some tests and if I include hyperloop code into the the `assets/` folder of the module I can build the module without problems. I have also added `hyperloop` module into the `timodule.xml`.
However when i use the new module from an app, I got errors related to native requires:

```
[ERROR] TiExceptionHandler: Error: Requested module not found: android.app.Activity
[ERROR] TiExceptionHandler:     at Module.require (ti:/module.js:297:8)
```

If I add the native requires into the `app.js` itself the module works fine.
I think that during the build, hyperloop scans the source code for any native requires, but I doesn't scan the module code (actually it should be already obfuscate).
I think a similar process (generating metabase) should be done during module building too.

thank you

**Reply by jquick-axway** [reply]
Right, hyperloop will only scan the app project's JS scripts. Hyperloop will ignore all module code. This is by design. No one has ever asked for hyperloop to operate on modules before.

So, one way to do it is to just distribute the hyperloop JS file and private AAR and tell app developers to copy them into their app project.

Alternatively, you can make a native module. The advantage of this is you won't punish app developers with the build overhead hyperloop adds to the build time.

If your private AAR is hosted on online, then doing this via a native module will be "easy" because you could reference it as a dependency via the module's "build.gradle". This is how Google wants you to do it.

If your private AAR is not hosted online (ie: your distributing it as a local file), then it's a bit more problematic. Google's build system will "link" to the local AAR when building a library, but it won't merge the local AAR into the AAR you're building. This is a limitation on Google's end. Google's build system will only merge AARs when building an app. So, this means the app needs a copy of the private AAR in the `./platform/android` folder too (the same thing you have to do for hyperloop).
Alternatively, you can unzip your private AAR and place its JAR and res files into your module's folders and those will be merged into your module's AAR that you're building.

Another approach might be to put the private Java/Kotlin classes in your native module and then obfuscate it with proguard.
